Source/missiles.cpp

```diff
@@ -506,38 +500,65 @@ void MoveMissileAndCheckMissileCol(Missile &missile, int mindam, int maxdam, boo
 			if (tile == missile.position.tile)
 				break;
 
-			// don't call CheckMissileCol more than once for a tile
+			// don't call checkTile more than once for a tile
 			if (prevTile == tile)
 				continue;
-			prevTile = tile;
-
-			CheckMissileCol(missile, mindam, maxdam, false, tile, false);
 
-			// Did missile hit anything?
-			if (missile._mirange != 0)
-				continue;
+			prevTile = tile;
 
-			if ((missile._miHitFlag && MissilesData[missile._mitype].MovementDistribution == MissileMovementDistrubution::Blockable) || IsMissileBlockedByTile(tile)) {
+			if (!checkTile(tile)) {
 				missile.position.traveled = traveled;
-				if (ifCollidesDontMoveToHitTile && missile._mirange == 0) {
+				if (ifCheckTileFailsDontMoveToTile) {
 					missile.position.traveled -= incVelocity;
 					UpdateMissilePos(missile);
 					missile.position.StopMissile();
 				} else {
 					UpdateMissilePos(missile);
 				}
-				return;
+				return true;
 			}
+
 		} while (true);
 	}
-	if (ignoreStart && missile.position.start == missile.position.tile)
-		return;
-	CheckMissileCol(missile, mindam, maxdam, false, missile.position.tile, false);
-	if (ifCollidesDontMoveToHitTile && missile._mirange == 0) {
+
+	if (!checkTile(missile.position.tile) && ifCheckTileFailsDontMoveToTile) {
 		missile.position.traveled -= missile.position.velocity;
 		UpdateMissilePos(missile);
 		missile.position.StopMissile();
 	}
+
+	return true;
+}
+
+void MoveMissileAndCheckMissileCol(Missile &missile, int mindam, int maxdam, bool ignoreStart, bool ifCollidesDontMoveToHitTile)
+{
+	auto checkTile = [&](Point tile) {
+		if (ignoreStart && missile.position.start == tile)
+			return true;
+
+		CheckMissileCol(missile, mindam, maxdam, false, tile, false);
+
+		// Did missile hit anything?
+		if (missile._mirange != 0)
+			return true;
+
+		if (missile._miHitFlag && MissilesData[missile._mitype].MovementDistribution == MissileMovementDistrubution::Blockable)

```

Comment:
It's not relevant to this PR, but I just noticed the typo in MissileMovementDistr<ins>**u**</ins>bution.

Reply:
Sorry for that one. 😳 
I fixed it in a separate commit.
if you think this should be better in another pr, I can drop the commit.